### PR TITLE
add fee and utxo handling, order serialization and signing

### PR DESCRIPTION
**Pull request by buck54321**
- Update fee information. Add fee calculation equations. 
- Add serialization instructions for all order types. 
- Add match notification structures
- Add progress report structures
- Explain special handling of market buy orders
- Change API call for lot size to more comprehensive asset info structure
- Add sections for UTXO selection and order signing

### CODE REVIEW

File: spec.mediawiki [hunk 1]
```diff
@@ -202,50 +204,87 @@ This section describes each of these design aspects.
...
+Instead, a one-time fee is collected by the pool during registration.
+Registration fees discourage certain spam attacks, enable punitive actions when 
+[[#Rules_of_Community_Conduct|conduct rules]] are violated, and help to cover 
+DEX operating expenses.
+Registration fees will be configurable by the exchange operator.
+
+====Transaction Fees====
+
+The clients will cover on-chain transaction fees at a fee rate set by the DEX 
+operator.
```
Comment (chappjc) [inline spec.mediawiki:223]: "at a minimum rate"

File: spec.mediawiki [hunk 2]
```diff
@@ -202,50 +204,87 @@ This section describes each of these design aspects.
...
+Registration fees will be configurable by the exchange operator.
+
+====Transaction Fees====
+
+The clients will cover on-chain transaction fees at a fee rate set by the DEX 
+operator. 
+Failure to adhere to the specified fee rate will result in a conduct violation.
+As part of any submitted order a client is required to demonstrate control of
+the funds for both the order itself, and the maximum fees that could be 
+assessed.
```
Comment (chappjc) [inline spec.mediawiki:228]: I might rephrase this to something like,
"As part of any submitted order, a client is required to demonstrate control of funds that will back the atomic swap, and ensure that the backing funds are sufficient to create the swap contract transaction with the necessary contract amount while covering the network's transaction fees at the specified minimum rate."
Maybe that's a bit long, but I think it's helpful to spell out what the backing funds cover and that the fee is a function of the network and the actual swap transaction size.

File: spec.mediawiki [hunk 3]
```diff
@@ -202,50 +204,87 @@ This section describes each of these design aspects.
...
+====Transaction Fees====
+
+The clients will cover on-chain transaction fees at a fee rate set by the DEX 
+operator. 
+Failure to adhere to the specified fee rate will result in a conduct violation.
+As part of any submitted order a client is required to demonstrate control of
+the funds for both the order itself, and the maximum fees that could be 
+assessed. 
+
+Some on-chain fees are higher than others. Large order will likely be filled
```
Comment (chappjc) [inline spec.mediawiki]: "Large order**s**" or "A large order"

Comment (chappjc) [inline spec.mediawiki]: Consider being more explicit with "Some on-chain fees are higher than others", instead explaining like "Total on-chain fees associated with an order will increase as the number of swap transactions required to settle the order increases.  For example, ..."

File: spec.mediawiki [hunk 4]
```diff
@@ -202,50 +204,87 @@ This section describes each of these design aspects.
...
+As part of any submitted order a client is required to demonstrate control of
+the funds for both the order itself, and the maximum fees that could be 
+assessed. 
+
+Some on-chain fees are higher than others. Large order will likely be filled 
+with several matches, requiring more transactions and therefore more fees. 
+The DEX operator must find a balance between low fees, which are preferable from 
+an accounting standpoint, or higher fees, which can decrease settlement time by 
+reducing the number of blocks before the transaction is mined. For most assets, 
+very low fees are likely possible without increasing the time until mined.
```
Comment (chappjc) [inline spec.mediawiki]: I'd personally prefer "For many assets, low fees are possible without..."  And if I were feeling devious I might end it with "..., although assets pushing the limits of on-chain scaling may have high fees.".

File: spec.mediawiki [hunk 5]
```diff
@@ -202,50 +204,87 @@ This section describes each of these design aspects.
...
+an accounting standpoint, or higher fees, which can decrease settlement time by 
+reducing the number of blocks before the transaction is mined. For most assets, 
+very low fees are likely possible without increasing the time until mined. 
+
+===Asset Variables===
+
+There are a number of asset-specific variables that must be known by the client.
+
+The '''lot size''' for an asset serves as both the minimum order quantity and 
+the order quantity increment. In particular, for ''lot size''  '''''l''''', the
```
Comment (chappjc) [inline spec.mediawiki]: If possible, I think it would he helpful to describe this in terms of a market's base asset (or quote asset if that's every the case).

File: spec.mediawiki (hunk 1 above)
Comment (buck54321) [reply spec.mediawiki:223]: Not sure on this one. In some sense, the client is expected to pay exactly the specified rate since order validation assumes it and order completion may depend on it.

Comment (chappjc) [reply spec.mediawiki:223]: Paying an exact rate reliably isn't really possible though because it's a fraction of atoms/bytes, integers where the bytes can't be controlled.

Even with dcrwallet set to use the standard 0.0001 DCR/kB, it rarely ends up that way.  Take for example these two transactions, each with two inputs and two outputs:
 - size: 417 bytes, fee: 4190 atoms, fee rate: 0.00010047 DCR/kB, https://explorer.dcrdata.org/tx/545ba9eb49b479d66e20cc025b1d106ec6a0ec168b23e69eaaa08db08ec21115
- size: 416 bytes, fee: 4190 atoms, fee rate: 0.00010072 DCR/kB, https://explorer.dcrdata.org/tx/ac345f94f6b43d039a83f6a0d699156723a2812f24261432579285a37757834d

Neither is exactly 0.0001 DCR/kB, and they aren't the same size anyway.

**Review by chappjc — APPROVED** [review]
